flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/sink/writer/MongoWriter.java

```diff
@@ -0,0 +1,183 @@
+
+        SinkWriterMetricGroup metricGroup = checkNotNull(initContext.metricGroup());
+        metricGroup.setCurrentSendTimeGauge(() -> ackTime - lastSendTime);
+
+        this.collector =
+                new CountingCollector<>(
+                        new ListCollector<>(this.bulkRequests),
+                        metricGroup.getNumRecordsSendCounter());
+
+        // Initialize the serialization schema.
+        this.sinkContext = new MongoSinkContextImpl(initContext, writeOptions);
+        try {
+            SerializationSchema.InitializationContext initializationContext =
+                    initContext.asSerializationSchemaInitializationContext();
+            serializationSchema.open(initializationContext, sinkContext, writeOptions);
+        } catch (Exception e) {
+            throw new FlinkRuntimeException("Failed to open the MongoEmitter", e);
+        }
+
+        // Initialize the mongo client.
+        this.mongoClient = MongoClients.create(connectionOptions.getUri());
+    }
+
+    @Override
+    public void write(IN element, Context context) throws IOException, InterruptedException {
+        // do not allow new bulk writes until all actions are flushed
+        while (checkpointInProgress) {
+            mailboxExecutor.yield();
+        }
+        collector.collect(serializationSchema.serialize(element, sinkContext));
+        if (isOverMaxActionsLimit() || isOverMaxFlushIntervalLimit()) {
+            doBulkWrite();
+        }
+    }
+
+    @Override
+    public void flush(boolean endOfInput) throws IOException {
+        checkpointInProgress = true;
+        while (!bulkRequests.isEmpty() && (flushOnCheckpoint || endOfInput)) {
+            doBulkWrite();
+        }
+        checkpointInProgress = false;
+    }
+
+    @Override
+    public void close() {
+        if (mongoClient != null) {
+            mongoClient.close();
+        }
+    }
+
+    @VisibleForTesting
+    void doBulkWrite() throws IOException {
+        if (bulkRequests.isEmpty()) {
+            // no records to write
+            return;
+        }
+
+        int maxRetryTimes = writeOptions.getMaxRetryTimes();
+        for (int i = 0; i <= maxRetryTimes; i++) {
+            try {
+                lastSendTime = System.currentTimeMillis();
+                mongoClient
+                        .getDatabase(connectionOptions.getDatabase())
+                        .getCollection(connectionOptions.getCollection(), BsonDocument.class)
+                        .bulkWrite(bulkRequests);
+                ackTime = System.currentTimeMillis();
+                bulkRequests.clear();
+                break;
+            } catch (MongoException e) {
+                LOG.error("Bulk Write to MongoDB failed, retry times = {}", i, e);
+                if (i >= maxRetryTimes) {
+                    throw new IOException(e);
+                }
+                try {
+                    Thread.sleep(1000L * i);
```

Comment:
this should be configurable.

The first attempt also doesn't wait at all which is likely not intended.

Reply:
Add the configuration `sink.retry.interval` and fixed the sleep time for the first attempt.